sql/time_bucket_ng.sql

```diff
@@ -24,20 +24,52 @@
-	AS '@MODULE_PATHNAME@', 'ts_time_bucket_ng_date' LANGUAGE C IMMUTABLE PARALLEL SAFE STRICT;
+    AS '@MODULE_PATHNAME@', 'ts_time_bucket_ng_date' LANGUAGE C IMMUTABLE PARALLEL SAFE STRICT;
 
 -- utility functions
 CREATE OR REPLACE FUNCTION timescaledb_experimental.time_bucket_ng(bucket_width INTERVAL, ts TIMESTAMP) RETURNS TIMESTAMP
-	AS '@MODULE_PATHNAME@', 'ts_time_bucket_ng_timestamp' LANGUAGE C IMMUTABLE PARALLEL SAFE STRICT;
+    AS '@MODULE_PATHNAME@', 'ts_time_bucket_ng_timestamp' LANGUAGE C IMMUTABLE PARALLEL SAFE STRICT;
 
 CREATE OR REPLACE FUNCTION timescaledb_experimental.time_bucket_ng(bucket_width INTERVAL, ts TIMESTAMP, origin TIMESTAMP) RETURNS TIMESTAMP
-	AS '@MODULE_PATHNAME@', 'ts_time_bucket_ng_timestamp' LANGUAGE C IMMUTABLE PARALLEL SAFE STRICT;
+    AS '@MODULE_PATHNAME@', 'ts_time_bucket_ng_timestamp' LANGUAGE C IMMUTABLE PARALLEL SAFE STRICT;
 
+-- The following two versions of time_bucket_ng() are kept for the backward
+-- compatibility with time_bucket(). Basically they treat 'ts' as a UTC timestamp
+-- (because it's its internal representation) which is almost certainly not
+-- something you want. Future versions may WARN you about this fact, and be
+-- completely removed eventually. Please use the procedures with 'timezone' TEXT
+-- argument instead.
 CREATE OR REPLACE FUNCTION timescaledb_experimental.time_bucket_ng(bucket_width INTERVAL, ts TIMESTAMPTZ) RETURNS TIMESTAMPTZ
-	AS '@MODULE_PATHNAME@', 'ts_time_bucket_ng_timestamptz' LANGUAGE C STABLE PARALLEL SAFE STRICT;
+    AS '@MODULE_PATHNAME@', 'ts_time_bucket_ng_timestamptz' LANGUAGE C STABLE PARALLEL SAFE STRICT;
 
 CREATE OR REPLACE FUNCTION timescaledb_experimental.time_bucket_ng(bucket_width INTERVAL, ts TIMESTAMPTZ, origin TIMESTAMPTZ) RETURNS TIMESTAMPTZ
-	AS '@MODULE_PATHNAME@', 'ts_time_bucket_ng_timestamptz' LANGUAGE C STABLE PARALLEL SAFE STRICT;
+    AS '@MODULE_PATHNAME@', 'ts_time_bucket_ng_timestamptz' LANGUAGE C STABLE PARALLEL SAFE STRICT;
+
+-- TIMESTAMPTZ version of time_bucket_ng().
+--
+-- It's a bit difficult to think about timezones, thus before you hastily
+-- conclude that something is wrong with the code, please consider the fact
+-- that we have tests that verify it.
+--
+-- Also, before you conclude that the code is slow because it's written in
+-- PL/pgSQL please do an actual benchmark. The benchmarks we did showed no
+-- measurable penalty for using PL/pgSQL for a simple case like this one.
+CREATE OR REPLACE FUNCTION timescaledb_experimental.time_bucket_ng(bucket_width INTERVAL, ts TIMESTAMPTZ, timezone TEXT) RETURNS TIMESTAMPTZ AS $$
+DECLARE
+    ret TIMESTAMP;
+BEGIN
+    -- Convert 'ts' to TIMESTAMP at given 'timezone', then treat resulting
+    -- timestamp as a regular one.
+    SELECT timescaledb_experimental.time_bucket_ng(
+             bucket_width,
+             ts at time zone timezone) INTO ret;
```

Comment:
I think we need both functions to have a good user experience is more complicated situations, but I would like to understand if this works in an equivalent way, if there is something preventing this use-case, or if I have misunderstood something.

Reply:
Isn't the above error something we can fix rather than being forced to have overloaded functions? I do realize that we need the `timestamptz` overload for other cases, but we should probably try to support this use case as well for a good user experience.